### PR TITLE
8266821: G1: Prefetch cards during merge heap roots phase

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that improves performance of the merge heap roots phase in g1?

Merging heap roots (remembered sets and log buffers) into the card table for later scanning to a large degree constitutes random accesses to the card table.

Putting a small prefetch queue between calculating the card table address and inspecting the given card table value decreases merge remembered set time by 20-30% and merge log buffers time by 40-50% (on x64. AArch64 shows similar if not better improvements).

Applications not having a significant amount of either remembered sets or log buffers do not show significant difference. 

The cache size constants for arm64 and x86 were derived from testing with a benchmark creating lots of references and log buffers (i.e. BigRamTester). Could not do any corresponding tests on other architectures due to lack of machines.

Testing: tier1-3, lots of tier1-tier8 runs with [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266821](https://bugs.openjdk.java.net/browse/JDK-8266821): G1: Prefetch cards during merge heap roots phase

### Reviewers
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to 288ba6dce5327c310b3232b3f2c0176162d0d1ca
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer) ⚠️ Review applies to 288ba6dce5327c310b3232b3f2c0176162d0d1ca

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3975/head:pull/3975` \
`$ git checkout pull/3975`

Update a local copy of the PR: \
`$ git checkout pull/3975` \
`$ git pull https://git.openjdk.java.net/jdk pull/3975/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3975`

View PR using the GUI difftool: \
`$ git pr show -t 3975`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3975.diff">https://git.openjdk.java.net/jdk/pull/3975.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/arm/gc/g1/g1Globals_arm.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef CPU_AARCH64_GC_G1_G1GLOBALS_ARM_HPP
```
Comment (kimbarrett) [inline src/hotspot/cpu/arm/gc/g1/g1Globals_arm.hpp:24]: s/AARCH64/ARM/ in 3 places in this file.  Similar changes in all the other non-aarch64 files.

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 2]
```diff
@@ -1060,10 +1061,61 @@ void G1RemSet::prepare_for_scan_heap_roots() {
...
+
+  uint _cur_cache_idx;
+
+  NONCOPYABLE(G1MergeHeapRootsPrefetchCache);
+
+protected:
+  // Initial content of all elements in the cache. It's value be "neutral", i.e.
+  // no work done on it.
+  G1CardTable::CardValue _dummy_card;
+
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1084]: This class should have a protected default destructor.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/arm/gc/g1/g1Globals_arm.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/cpu/arm/gc/g1/g1Globals_arm.hpp]: Sorry, copy&paste :(

**Review by walulyai — APPROVED** [review]
lgtm!

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 3]
```diff
@@ -1078,10 +1078,12 @@ class G1MergeHeapRootsPrefetchCache {
   NONCOPYABLE(G1MergeHeapRootsPrefetchCache);
 
 protected:
-  // Initial content of all elements in the cache. It's value be "neutral", i.e.
-  // no work done on it.
+  // Initial content of all elements in the cache. It's value should be
+  // "neutral", i.e. no work done on it when processing it.
   G1CardTable::CardValue _dummy_card;
 
+  ~G1MergeHeapRootsPrefetchCache() { }
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1085]: Better would be a defaulted definition, i.e. `~G1MergeHeapRootsPrefetchCache() = default;`

**Review by kimbarrett — APPROVED** [review]
One minor nit, otherwise looks good.

File: src/hotspot/share/gc/g1/g1RemSet.cpp [hunk 4]
```diff
@@ -1060,10 +1061,63 @@ void G1RemSet::prepare_for_scan_heap_roots() {
...
+  uint _cur_cache_idx;
+
+  NONCOPYABLE(G1MergeHeapRootsPrefetchCache);
+
+protected:
+  // Initial content of all elements in the cache. It's value should be
+  // "neutral", i.e. no work done on it when processing it.
+  G1CardTable::CardValue _dummy_card;
+
+  ~G1MergeHeapRootsPrefetchCache() { }
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1085]:
```suggestion
  ~G1MergeHeapRootsPrefetchCache() = default;
```

**Comment by tschatzl** [comment]
Thanks @walulyai @kimbarrett for your reviews.

/integrate

**Pull request closed without merging**